packages/compat/src/transaction.ts

```diff
@@ -0,0 +1,144 @@
+import {
+    appendTransactionInstruction,
+    Blockhash,
+    createTransaction,
+    ITransactionWithBlockhashLifetime,
+    ITransactionWithFeePayer,
+    ITransactionWithSignatures,
+    setTransactionFeePayer,
+    setTransactionLifetimeUsingBlockhash,
+    Transaction,
+} from '@solana/transactions';
+import {
+    MessageAccountKeys,
+    MessageCompiledInstruction,
+    PublicKey,
+    VersionedMessage,
+    VersionedTransaction,
+} from '@solana/web3.js';
+
+import { SignedVersionedTransaction } from './signed-versioned-transaction';
+
+function convertAccount(
+    message: VersionedMessage,
+    accountKeys: MessageAccountKeys,
+    accountIndex: number
+): IAccountMeta {
+    const accountPublicKey = accountKeys.get(accountIndex);
+    if (!accountPublicKey) {
+        // TODO coded error
+        throw new Error(`Could not find account address at index ${accountIndex}`);
+    }
+    const isSigner = message.isAccountSigner(accountIndex);
+    const isWritable = message.isAccountWritable(accountIndex);
+
+    const role = isSigner
+        ? isWritable
+            ? AccountRole.WRITABLE_SIGNER
+            : AccountRole.READONLY_SIGNER
+        : isWritable
+        ? AccountRole.WRITABLE
+        : AccountRole.READONLY;
+
+    return {
+        address: accountPublicKey.toBase58() as Base58EncodedAddress,
+        role,
+    };
+}
+
+function convertInstruction(
+    message: VersionedMessage,
+    accountKeys: MessageAccountKeys,
+    instruction: MessageCompiledInstruction
+): IInstruction {
+    const programAddressPublicKey = accountKeys.get(instruction.programIdIndex);
+    if (!programAddressPublicKey) {
+        // TODO coded error
+        throw new Error(`Could not find program address at index ${instruction.programIdIndex}`);
+    }
+
+    const accounts = instruction.accountKeyIndexes.map(accountIndex =>
+        convertAccount(message, accountKeys, accountIndex)
+    );
+
+    return {
+        programAddress: programAddressPublicKey.toBase58() as Base58EncodedAddress,
+        ...(accounts.length ? { accounts } : {}),
+        ...(instruction.data.length ? { data: instruction.data } : {}),
+    };
+}
+
+function convertSignatures(
+    transaction: VersionedTransaction,
+    staticAccountKeys: PublicKey[]
+): ITransactionWithSignatures['signatures'] {
+    return transaction.signatures.reduce((acc, sig, index) => {
+        // legacy web3js includes a fake all 0 signature if it hasn't been signed
+        // we don't do that for the new tx model. So just skip if it's all 0s
+        const allZeros = sig.every(byte => byte === 0);
+        if (allZeros) return acc;
+
+        const address = staticAccountKeys[index].toBase58() as Base58EncodedAddress;
+        return { ...acc, [address]: sig as Ed25519Signature };
+    }, {});
+}
+
+export function fromVersionedTransactionWithBlockhash(
+    transaction: SignedVersionedTransaction,
+    blockhash: Blockhash,
+    lastValidBlockHeight: bigint
```

Comment:
Previously I used the last blockhash from their transaction but required them to pass `lastValidBlockHeight`. Then I refactored to have them pass both, since they'd get both together. If just defaulting `lastValidBlockHeight` to `U64_MAX` works for us then I'm happy to change to doing that, for some reason I thought you had to use the value that matches what you got when you fetched the blockhash. Will update, thanks!

Reply:
yeah I didn't know you could do this tbh

Reply:
The function of `lastValidBlockHeight` is to signal to the confirmation function when it’s time to give up. As in: hey, if the slot is higher than X, stop waiting for the transaction to confirm.

Reply:
Defaulting this to some very high number essentially disables this feature. That isn’t great either, and come to think of it is a behavioural change from what happens now if you don’t supply a block height (ie. we timeout after X seconds).

Sigh.

Options:

1. We create first class support in modern for not specifying a block height (please no)
2. We make you supply a block height when you `fromVersionedTransaction`; halfway between this and what you had before.
3. We make the confirmation function unconditionally time out after X, meaning that all a block height could possibly do would be to make it time out sooner (pretty much the intent of it)
4. Sort of like 3 crossed with 1, make a unique `Symbol` that’s acceptable in the type position for `lastValidBlockHeight` that represents `UNKNOWN_BLOCK_HEIGHT` and use that.

Reply:
> 1. We create first class support in modern for not specifying a block height (please no)

Agreed, definitely don't want to change experimental behavior as a result of how legacy is written

2 is probably the most functionally correct, but the dev ex is rough. Why don't we implement the behavior of 3, but add an overload for providing a block height instead of X? That way, if you convert over and don't pass it, you get default X, if you pass it _or_ if you're doing send/confirm in the new lib without `compat`, you get the proper control?

Reply:
I think I reiterated 3 more than I realized, but emphasis on the overloading

Reply:
I think that's probably the best option? Make `lastValidBlockHeight` optional, default it to some extremely large number that effectively makes it useless.

The only question is whether to ship a version of `confirmTransaction` that:

* times out after some fixed number of seconds
* never times out. no time out. no more.

I'm sort of inclined to eliminate fixed timeouts completely. This means that if you convert a legacy transaction, and _don't_ supply a `lastValidBlockHeight`, your confirmations will hang indefinitely in the event that the transactions fail. Timing out is up to you at that point.

Actually, I kind of like that. Thoughts?

Reply:
Wouldn't this take up a thread on your app though? What if you send a bunch of transactions and they are all waiting for confirmation and just keep one or more subscriptions open? Might be bad for performance. I also am not sure how you'd go about force-cancelling that

Reply:
Not really a thread, it would just hang on a continuation that will never continue. It would keep pumping `slotSubscribe()` though, for no reason. If your app gets tired of waiting it can cancel it with the `AbortSignal` you passed in.

I think it would look something like this:

```ts
// MyApp

// Missing sensible `lastValidBlockHeight` so this will try to confirm forever.
const tx = fromVersionedTransaction(legacyTransaction);
const abortController = new AbortController();
const confirmationPromise = confirmTransaction({abortSignal: abortController.signal}, tx);
const timeoutPromise = new Promise((_, reject) => {
  AbortSignal
    .timeout(60_000)
    .addEventListener('abort', reject, {signal: abortController.signal});
});
try {
  await Promise.race([
    confirmationPromise,
    timeoutPromise,
  ]);
  console.log('confirmed');
} catch(e) {
  console.error('o no', e);
} finally {
  abortController.abort();
}
```

Reply:
Updated to have an optional `lastValidBlockHeight` and default to U64 max